vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -448,10 +448,17 @@ and
                                                 optionsAssociation.Remove(projectContext) |> ignore
                                                 project.Disconnect()))
                 for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
-                    let referencedProjectId = setup referencedSite                    
+                    let referencedProjectFileName = referencedSite.ProjectFileName()
+                    let referencedProjectDisplayName = projectDisplayNameOf referencedProjectFileName
+                    let referencedProjectId = workspace.ProjectTracker.GetOrCreateProjectIdForPath(referencedProjectFileName, referencedProjectDisplayName)
                     project.AddProjectReference(ProjectReference referencedProjectId)
-                workspace.ProjectTracker.AddProject(project)
-            projectId
+
+                if not (workspace.ProjectTracker.ContainsProject(project)) then 
+                    workspace.ProjectTracker.AddProject(project)
+
+                for referencedSite in ProjectSitesAndFiles.GetReferencedProjectSites (site, this.SystemServiceProvider) do
```

Comment:
For perf, I would suggest adding the reference *before* telling the workspace about the project.  That would avoid sending a bunch of `ProjectChanged` workspace notifications that will cancel and kick off new work.